docs/docs/deployment/docker.md

```diff
@@ -0,0 +1,109 @@
+:::
+
+Exograph provides three Docker images:
+
+- `ghcr.io/exograph/cli`: It includes only the [`exo` CLI](../cli-reference/development) and may be used for development tasks such as building the `exo_ir` file and computing the database schema. Note that you cannot run `exo yolo` or `exo dev` using this image (you can use `ghcr.io/exograph/dev` for this purpose).
+- `ghcr.io/exograph/server`: It includes only the Exograph server and forms the basis for running it in production.
+- `ghcr.io/exograph/dev`: It includes the Exograph server, the `exo` CLI, and Postgres 16. Since it is optimized for development, it is a rather large image, and it is **not recommended** for production use. Typically, you will enter [into a shell for this image to perform development tasks, including `exo yolo` and `exo dev`](../getting-started/docker).
+
+Let's see how you can use these images for specific tasks.
+
+## Building the `exo_ir` file
+
+To run an Exograph server, you must build an `exo_ir` file. While you would typically build this file using the [build](../cli-reference/development/build.md) command to use the locally installed `exo` CLI, you can also build this file using a Docker container using the `ghcr.io/exograph/cli` image (you may also use `ghcr.io/exograph/dev`).
+
+```sh
+# shell-command-next-line
+docker run --rm --platform linux/x86_64 \
+  --mount type=bind,source="$(pwd)",target=/usr/src/app \
+  -it ghcr.io/exograph/cli:latest bash -c "exo build"
+```
+
+This will build the `exo_ir` file and put it in the `target` directory.
+
+## Running the server
+
+To run the server locally, you will need to pass the Postgres connection URL such that the Docker container can resolve it.
+
+```sh
+# shell-command-next-line
+docker run --rm --platform linux/x86_64 \
+  --mount type=bind,source="$(pwd)",target=/usr/src/app \
+  -e EXO_POSTGRES_URL=<postgres-url> \
+  -e EXO_INTROSPECTION=true \
+  -p 9876:9876 -it ghcr.io/exograph/dev:latest exo-server
+```
+
+The URL form will depend on the database you are using.
+
+- **Cloud database**: If you use a cloud Postgres database such as [Neon](https://neon.tech/), simply pass the connection URL given to you by the database provider.
+- **Local database**: If you use a local Postgres database, you may use `host.docker.internal` as the host since this is a special DNS name that resolves to the host machine. That URL will be of the form `postgres://<user>:<password>@host.docker.internal:5432/<database>`.
+
+Of course, you may pass any additional environment such as `EXO_INTROSPECTION`.
+
+## Building a Docker image
+
+If you need a generic (cloud-provider-agnostic) Docker image for your application, you can start with the following Dockerfile:
+
+```Dockerfile
+FROM ghcr.io/exograph/cli:latest as builder
+
+WORKDIR /app
+
+COPY ./src ./src
+
+RUN exo build
```

Comment:
I had to add: `USER root` otherwise there was a permission error during build

Reply:
Can you tell me more about this (which version of Docker, etc.)? I tried to rebuild, and I got no error without `USER root`. While it may be okay to add `USER root` here, ideally, we can work with least privileges.

Reply:
Docker version:

```
 Version:    25.0.4-rd
 Context:    desktop-linux
```

```
> [exograph_prod builder 4/4] RUN exo build:
0.225 error: Permission denied (os error 13)
0.225 
0.236 Error: Parser error: Permission denied (os error 13)
0.236 
```

Reply:
Interesting. I tried from Mac (Docker version 25.0.3, build 4debf41) and Linux (Docker version 25.0.6, build v25.0.6) and still no error. Let's see if we can figure this out.

Can you:
- Add `RUN ls -l ./src` above `exo build`
- Run `docker build --no-cache --progress=plain -t todo-api .`

I suspect there is some difference in how the `src` got copied.

Reply:
```
❯ docker build --no-cache --progress=plain -t todo-api -f ./services/exograph/Dockerfile .
#0 building with "desktop-linux" instance using docker driver

#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 818B done
#1 DONE 0.0s

#2 resolve image config for docker-image://docker.io/docker/dockerfile:1.10
#2 DONE 0.5s

#3 docker-image://docker.io/docker/dockerfile:1.10@sha256:865e5dd094beca432e8c0a1d5e1c465db5f998dca4e439981029b3b81fb39ed5
#3 CACHED

#4 [internal] load metadata for ghcr.io/exograph/server:v0.9.9
#4 ...

#5 [internal] load metadata for ghcr.io/exograph/cli:v0.9.9
#5 DONE 0.2s

#4 [internal] load metadata for ghcr.io/exograph/server:v0.9.9
#4 DONE 0.2s

#6 [internal] load .dockerignore
#6 transferring context: 198B done
#6 DONE 0.0s

#7 [builder 1/5] FROM ghcr.io/exograph/cli:v0.9.9@sha256:b1f5669aa0dd2d4ccc9262f859041565a9ea8ca8258395555d60b8529c1a5e5f
#7 resolve ghcr.io/exograph/cli:v0.9.9@sha256:b1f5669aa0dd2d4ccc9262f859041565a9ea8ca8258395555d60b8529c1a5e5f done
#7 DONE 0.0s

#8 [stage-1 1/3] FROM ghcr.io/exograph/server:v0.9.9@sha256:8facc5a2ca9ffd04a314a7821367f61f99332d3cdd7972ee359dc5d67b04a0d4
#8 resolve ghcr.io/exograph/server:v0.9.9@sha256:8facc5a2ca9ffd04a314a7821367f61f99332d3cdd7972ee359dc5d67b04a0d4 done
#8 DONE 0.0s

#9 [builder 2/5] WORKDIR /app
#9 CACHED

#10 [stage-1 2/3] WORKDIR /app
#10 CACHED

#11 [internal] load build context
#11 transferring context: 428B done
#11 DONE 0.0s

#12 [builder 3/5] COPY ./concerts-app /app
#12 DONE 0.0s

#13 [builder 4/5] RUN ls -l ./src
#13 0.111 total 12
#13 0.111 -rw-r--r-- 1 root root  992 Oct  3 18:36 index.exo
#13 0.111 -rw-r--r-- 1 root root 1009 Oct  3 18:35 notification.ts
#13 0.111 -rw-r--r-- 1 root root  229 Oct  3 18:35 time.js
#13 DONE 0.1s

#14 [builder 5/5] RUN exo build
#14 0.212 error: Permission denied (os error 13)
#14 0.212 
#14 0.225 Error: Parser error: Permission denied (os error 13)
#14 0.225 
#14 ERROR: process "/bin/sh -c exo build" did not complete successfully: exit code: 1
------
 > [builder 5/5] RUN exo build:
0.212 error: Permission denied (os error 13)
0.212 
0.225 Error: Parser error: Permission denied (os error 13)
0.225 
------

 2 warnings found (use docker --debug to expand):
 - InvalidBaseImagePlatform: Base image ghcr.io/exograph/cli:v0.9.9 was pulled with platform "linux/amd64", expected "linux/arm64" for current build (line 2)
 - InvalidBaseImagePlatform: Base image ghcr.io/exograph/server:v0.9.9 was pulled with platform "linux/amd64", expected "linux/arm64" for current build (line 14)
Dockerfile:9
--------------------
   7 |     # USER root
   8 |     RUN ls -l ./src
   9 | >>> RUN exo build
  10 |     
  11 |     # You may add additional build steps such as
--------------------
ERROR: failed to solve: process "/bin/sh -c exo build" did not complete successfully: exit code: 1

View build details: docker-desktop://dashboard/build/desktop-linux/desktop-linux/lrl7c2pigfijix916hymrrxzv
~/development/promonow/promonow/prototyping/tech-exploration/exograph main ⇣1 !1 ❯ 
```

Reply:
`# syntax=docker/dockerfile:1.10` is the only difference I can spot on my end

Reply:
Thanks for exploring this. I am adding `USER root`. Since this is during the build step, implications are limited.

Later, I will see if I can reproduce it to understand this better.